test/AutoDiff/differentiable_func_type_parse.swift

```diff
@@ -0,0 +1,5 @@
+// RUN: %target-swift-frontend -parse -verify %s
+
+let a: @differentiable (Float) -> Float // okay
+
+let b: @differentiable(linear) (Float) -> Float // okay
```

Comment:
From what I saw, no parsing tests were done for marking types as differentiable so I created a new file.

Reply:
I am trying to add this test:
```swift
// expected-error @+2 {{expected ')' after 'linear' in '@differentiable' attribute}}
// expected-note @+1 {{to match this opening '('}}
let c: @differentiable(linear (Float) -> Float
```
But am getting an error:
```
differentiable_func_type_parse.swift:8:4: error: expected note not produced
// expected-note @+1 {{to match this opening '('}}
~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<unknown>:0: error: unexpected note produced: to match this opening '('
```

Reply:
Maybe the source location is wrong? Try running it in a terminal and see the source location.

Reply:
Ahh yes so it actually prints out differently then I was expecting. I included the second error below, however that is a type check error, not a parse error which this test doesn't see.

```swift
diff.swift:18:31: error: expected ')' after 'linear' in '@differentiable' attribute
let c: @differentiable(linear (Float) -> Float
                              ^
<unknown>:0: note: to match this opening '('
diff.swift:18:9: error: cannot mark types as linear differentiable
let c: @differentiable(linear (Float) -> Float
        ^
```

Reply:
It should not emit a type error when there's a parse error. It means that you probably didn't `return true` or something after you emitted a parser diagnostic.

Reply:
Yes now with the new way I handle the token with the backtrack so I can look ahead multiple tokens, I think this problem went away. Adding more test cases because of the change as well.